### ISSUE TITLE
Sync stop unfinished and the node start with 11% of the blockchain

### ISSUE THREAD

**Issue by M5oul** [duniter/duniter#521]
With Duniter Desktop 0.22.0 and a bad connection, which surely had been cut, I sync from zero my node, a timeout happen and the sync stopped and the node is launched:

![capture du 2016-08-10 17-57-24](https://cloud.githubusercontent.com/assets/4758871/17560985/615f1938-5f24-11e6-9527-c0f6d68d7d52.png)

```bash
2016-08-10T17:45:01+02:00 - debug: Peer 9bZEAT is on same blockchain
2016-08-10T17:45:01+02:00 - info: Getting chunck from 3000 to 3500
2016-08-10T17:45:02+02:00 - debug: 2500
2016-08-10T17:45:02+02:00 - info: Downloaded 8%, Applied 10%
2016-08-10T17:45:09+02:00 - info: Downloaded 10%, Applied 10%
2016-08-10T17:45:09+02:00 - debug: dl starts from 3500
2016-08-10T17:45:09+02:00 - debug: apply starts from 3000
2016-08-10T17:45:09+02:00 - debug: Applying from 3000
2016-08-10T17:45:09+02:00 - debug: Peer 9bZEAT is on same blockchain
2016-08-10T17:45:09+02:00 - info: Getting chunck from 3500 to 4000
2016-08-10T17:45:09+02:00 - debug: 3000
2016-08-10T17:45:09+02:00 - info: Downloaded 10%, Applied 11%
2016-08-10T17:45:11+02:00 - info: ETIMEDOUT
2016-08-10T17:45:11+02:00 - error: Error: ETIMEDOUT
    at Timeout._onTimeout (/home/moul/projects/duniter/sources/node_modules/vucoin/node_modules/request/request.js:761:15)
    at tryOnTimeout (timers.js:224:11)
    at Timer.listOnTimeout (timers.js:198:5)
2016-08-10T17:45:11+02:00 - warn: ETIMEDOUT
2016-08-10T17:45:11+02:00 - info: External access: 83.197.194.85:46608
2016-08-10T17:45:11+02:00 - debug: Generating server's peering entry based on block#469...
2016-08-10T17:45:11+02:00 - debug: Document write error:  Signature from a peer must match
2016-08-10T17:45:11+02:00 - error: TypeError: Cannot set property 'documentType' of undefined
    at /home/moul/projects/duniter/sources/app/service/PeeringService.js:281:27
    at next (native)
    at onFulfilled (/home/moul/projects/duniter/sources/node_modules/co/index.js:65:19)
    at process._tickCallback (internal/process/next_tick.js:103:7)
2016-08-10T17:45:11+02:00 - warn: Cannot set property 'documentType' of undefined
2016-08-10T17:52:55+02:00 - info: Duniter server listening on http://192.168.0.10:46608
2016-08-10T17:52:55+02:00 - info: Duniter server listening on http://::1:46608
2016-08-10T17:52:55+02:00 - info: Node version: 0.22.0
2016-08-10T17:52:55+02:00 - info: Node pubkey: AySm48XnZ1XqLJwhGtA5MqA24FGdvMDPkEg3zjARTXHc
2016-08-10T17:52:55+02:00 - info: Crawling the network...
2016-08-10T17:52:55+02:00 - warn: Local node is not a member. Waiting to be a member before computation a block.
2016-08-10T17:52:55+02:00 - info: Crawling done.
2016-08-10T17:52:55+02:00 - info: Storing self peer...
2016-08-10T17:52:55+02:00 - info: External access: 83.197.194.85:46608
2016-08-10T17:52:55+02:00 - debug: Generating server's peering entry based on block#469...
2016-08-10T17:52:55+02:00 - debug: Document write error:  Signature from a peer must match
2016-08-10T17:52:55+02:00 - error: TypeError: Cannot set property 'documentType' of undefined
    at /home/moul/projects/duniter/sources/app/service/PeeringService.js:281:27
    at next (native)
    at onFulfilled (/home/moul/projects/duniter/sources/node_modules/co/index.js:65:19)
    at process._tickCallback (internal/process/next_tick.js:103:7)
2016-08-10T17:52:55+02:00 - warn: Cannot set property 'documentType' of undefined
```

A solution could be to check all the blockchain is present before launching the node.
If not, display a message or retry sync.
Sync could be enhanced.

**Comment by M5oul** [comment]
It's a rare bug.
I could not reproduce it.
But, sync could be enhanced.

**Comment by M5oul** [comment]
I reproduced it syncthing from inso node with desktop duniter version 0.30.13:
```bash
2016-08-26T14:51:57+02:00 - debug: 4000
2016-08-26T14:51:57+02:00 - info: Downloaded 11%, Applied 13%
2016-08-26T14:51:58+02:00 - info: Downloaded 13%, Applied 13%
2016-08-26T14:51:58+02:00 - debug: dl starts from 5000
2016-08-26T14:51:58+02:00 - debug: apply starts from 4500
2016-08-26T14:51:58+02:00 - debug: Applying from 4500
2016-08-26T14:51:58+02:00 - debug: Peer 8Fi1VS is on same blockchain
2016-08-26T14:51:58+02:00 - info: Getting chunck from 5000 to 5500
2016-08-26T14:51:58+02:00 - trace: New node#10
2016-08-26T14:51:58+02:00 - trace: moricef was affected wotb_id 10
2016-08-26T14:51:58+02:00 - trace: New node#11
2016-08-26T14:51:58+02:00 - trace: jytou was affected wotb_id 11
2016-08-26T14:51:58+02:00 - trace: Link 0 -> 10
2016-08-26T14:51:58+02:00 - trace: Link 4 -> 10
2016-08-26T14:51:58+02:00 - trace: Link 6 -> 10
2016-08-26T14:51:58+02:00 - trace: Link 6 -> 11
2016-08-26T14:51:58+02:00 - trace: Link 0 -> 11
2016-08-26T14:51:58+02:00 - trace: Link 4 -> 11
2016-08-26T14:51:58+02:00 - debug: 4500
2016-08-26T14:51:59+02:00 - info: Downloaded 13%, Applied 14%
2016-08-26T14:52:00+02:00 - info: ETIMEDOUT
2016-08-26T14:52:00+02:00 - error: Error: ETIMEDOUT
    at Timeout._onTimeout (/home/moul/projects/duniter_build/sources/node_modules/vucoin/node_modules/request/request.js:761:15)
    at tryOnTimeout (timers.js:224:11)
    at Timer.listOnTimeout (timers.js:198:5)
2016-08-26T14:52:00+02:00 - warn: ETIMEDOUT
2016-08-26T14:52:00+02:00 - info: External access: 78.227.107.45:63748
2016-08-26T14:52:00+02:00 - debug: Generating server's peering entry based on block#469...
2016-08-26T14:52:00+02:00 - info: Next peering signal in 0 min
2016-08-26T14:55:56+02:00 - trace: UPnP: mapping external port 63748 to local 63748...
2016-08-26T15:00:56+02:00 - trace: UPnP: mapping external port 63748 to local 63748...
```

**Comment by M5oul** [comment]
Also, It seems the timeout value is not enough.
What is it's value?

**Comment by c-geek** [comment]
The value is given [here in `vucoinOptions`](https://github.com/duniter/duniter/blob/master/app/lib/sync.js#L94). The option is `vucoinOptions.timeout`.

If no value is given, default is `4000 ms`.

**Comment by c-geek** [comment]
Also, the algorithm needs 2 things:

* full download first: for an unknown reason, it is waiting for blocks application before downloading the next chunk
* parallel download on several nodes: this could be performed by first getting a list of nodes with the same current block, and download the blocks on these nodes in parallel to have a faster download. Of course an error handling with retries for a given chunk on another node is mandatory.

**Comment by c-geek** [comment]
Also a verification should be done about the good chaining of the blocks, since we no more have an algorithm where we can trust the peer, but an algorithm where: 

* we trust the current block of the peer
* the list of blocks should be chained correctly in both number & hash

**Comment by M5oul** [comment]
Nice :+1: I think it's the idea I wanted to express with [this ticket](https://github.com/duniter/duniter/issues/142) but with a big protocol :smile:

**Comment by c-geek** [comment]
Oh, indeed, it is the same idea :)

**Comment by M5oul** [comment]
I like the idea that we only need to take current block of trusted node, then download the blockchain from other nodes and verify it with the chained system without any trust issue :smile_cat:

**Comment by M5oul** [comment]
I tested the code, syncthing on `duniter.org:8999`  and get:
```bash
Progress:

Download: [                    ] 3 %
Apply:    [                    ] 3 %

Status: ETIMEDOUT
All done.
```

**Comment by M5oul** [comment]
Ok, understood, sync process timeout because of the upload bandwich was full.
I stopped uploads, and sync works well.

**Comment by c-geek** [comment]
I've not finished yet with the sync process, it was just a first commit.

**Comment by M5oul** [comment]
It still doesn't works.
```bash
Progress:

Download: [                    ] 0 %
Apply:    [                    ] 0 %

Status: ETIMEDOUT
All done.
```

May I open a new ticket?
